### PR TITLE
LPS-116762 User is not unsubscribed from Blogs portlet if leaves site through a user group

### PR DESCRIPTION
**Pull request by MarcellGyopos**
Hi @pei-jung ,
Could you please review this pull request?
Thank you and best regards,
Marci

### CODE REVIEW

**Comment by Sharryxue** [comment]
Trigger the CI test for a basic check befor starting the review.

**Comment by Sharryxue** [comment]
ci:test:sf

**Comment by Sharryxue** [comment]
ci:test:relevant

**Comment by Sharryxue** [comment]
Hello @MarcellGyopos ,
Would you please double checking the sf issue in UserLocalServiceImpl.java based on the SF result, thanks in advance.
/ cc @pei-jung

File: portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java [hunk 1]
```diff
@@ -4031,9 +4032,46 @@ public void unsetTeamUsers(long teamId, long[] userIds)
 	public void unsetUserGroupUsers(long userGroupId, long[] userIds)
 		throws PortalException {
 
-		userGroupPersistence.removeUsers(userGroupId, userIds);
+		long[] sites = userGroupLocalService.getGroupPrimaryKeys(userGroupId);
 
-		reindex(userIds);
+		for (long userId : userIds) {
+			Map<Long, long[]> userGroupGroupIds = new HashMap();
```
Comment (IstvanD) [inline portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java:4038]: The diamond operator is missing from the HashMap constructor. This should be
```
Map<Long, long[]> userGroupGroupIds = new HashMap<>();
```

File: portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java [hunk 2]
```diff
@@ -4031,9 +4032,46 @@ public void unsetTeamUsers(long teamId, long[] userIds)
...
+				if (ug.getUserGroupId() != userGroupId) {
+					userGroupGroupIds.put(
+						ug.getPrimaryKey(),
+						userGroupLocalService.getGroupPrimaryKeys(
+							ug.getUserGroupId()));
+				}
+			}
+
+			for (long s : sites) {
+				boolean isSiteContainsGroup = false;
```
Comment (IstvanD) [inline portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java:4053]: the variable should be named `siteContainsGroup`, not `isSiteContainsGroup`

File: portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java [hunk 3]
```diff
@@ -4031,9 +4032,46 @@ public void unsetTeamUsers(long teamId, long[] userIds)
...
+
+				for (Map.Entry<Long, long[]> entry :
+					userGroupGroupIds.entrySet()) {
```
Comment (IstvanD) [inline portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java:4056]: probably a tab is missing here

**Comment by MarcellGyopos** [comment]
Thank you @IstvanD :)

**Comment by MarcellGyopos** [comment]
Thank you @Sharryxue , I fixed it.

**Comment by MarcellGyopos** [comment]
ci:test:sf

**Comment by pei-jung** [comment]
ci:test:sf

**Comment by pei-jung** [comment]
Hi @MarcellGyopos 

This fix doesn't work if the user group is removed from the user via the Users Admin portlet (Control Panel -> Users and Organizations -> edit User -> Memberships). Since there're multiple ways to assign and unassign a user to user groups, a more robust approach will be to apply your fix in a model listener with either `onAfterRemoveAssociation` or `onBeforeRemoveAssociation`, whichever makes more sense.

It'd be nice if you can include a test with your fix to prevent future regressions. Also, i think since this issue is more related to site membership, it should go to review with the WEM(Echo) team. However, if they think this should be reviewed by USM, feel free to send your updated fix back to me.

Let me know if you have any questions. Thanks!

**Pull request closed without merging**